mjolnir/interaction/local/DirectionalContactInteraction.hpp

```diff
@@ -0,0 +1,390 @@
+    const coordinate_type& rCi = sys.position(Ci);
+    const coordinate_type& rPi = sys.position(Pi);
+    const coordinate_type& rPj = sys.position(Pj);
+    const coordinate_type& rCj = sys.position(Cj);
+
+    // =========================================================
+    // contact schema
+    //
+    //     theta1 theta2
+    //       |      |
+    //  Ci o v      v o Cj
+    //     \-.      ,-/
+    //   Pi o- - - - o Pj
+    //        |Pij|
+    //
+    // U_dir = U_angle1(theta1) * U_angle2(theta2) * U_contact(|Pij|)
+
+    const coordinate_type  Pij = sys.adjust_direction(rPj - rPi); // Pi -> Pj
+    const real_type       lPij = math::length(Pij);
+    if(lPij > contact_pot.cutoff())
+    {
+      continue;
+    }
+
+    // ==========================================================
+    // dU_angle1(theta1) / dr
+    const coordinate_type   PiCi        = sys.adjust_direction(rCi - rPi);
+    const real_type inv_len_PiCi        = math::rlength(PiCi);
+    const coordinate_type PiCi_reg      = PiCi * inv_len_PiCi;
+
+    const real_type inv_len_Pij         = real_type(1.0) / lPij;
+    const coordinate_type Pij_reg       = Pij * inv_len_Pij;
+
+    const real_type PiCi_dot_Pij        = math::dot_product(PiCi_reg, Pij_reg);
+    const real_type cos_theta1          = math::clamp(PiCi_dot_Pij,  real_type(-1.0), real_type(1.0));
+    const real_type theta1              = std::acos(cos_theta1);
+    const real_type angle1_coef         = angle1_pot.derivative(theta1);
+
+    const real_type sin_theta1          = std::sin(theta1);
+    const real_type angle1_coef_inv_sin = (sin_theta1 > math::abs_tolerance<real_type>()) ?
+        angle1_coef / sin_theta1 : angle1_coef / math::abs_tolerance<real_type>();
+
+    const coordinate_type dU_angle1_drCi =
+        (angle1_coef_inv_sin * inv_len_PiCi) * (cos_theta1 * PiCi_reg - Pij_reg);
+    const coordinate_type dU_angle1_drPj =
+        (angle1_coef_inv_sin * inv_len_Pij)  * (cos_theta1 * Pij_reg  - PiCi_reg);
+
+    const coordinate_type dU_angle1_drPi = -(dU_angle1_drCi + dU_angle1_drPj);
+
+    MJOLNIR_LOG_DEBUG("dU_angle1 / drCi = {x = ", math::X(dU_angle1_drCi), ", ",
+                      "y = ", math::Y(dU_angle1_drCi), ", ",
+                      "z = ", math::Z(dU_angle1_drCi), "}");
+    MJOLNIR_LOG_DEBUG("dU_angle1 / drPj = {x = ", math::X(dU_angle1_drPj), ", ",
+                      "y = ", math::Y(dU_angle1_drPj), ", ",
+                      "z = ", math::Z(dU_angle1_drPj), "}");
+    MJOLNIR_LOG_DEBUG("dU_angle1 / drPi = {x = ", math::X(dU_angle1_drPi), ", ",
+                      "y = ", math::Y(dU_angle1_drPi), ", ",
+                      "z = ", math::Z(dU_angle1_drPi), "}");
+
+    // dU_angle2(theta2) / dr
+    const coordinate_type PjCj          = sys.adjust_direction(rCj - rPj);
+    const real_type inv_len_PjCj        = math::rlength(PjCj);
+    const coordinate_type PjCj_reg      = PjCj * inv_len_PjCj;
+
+    const coordinate_type Pji_reg       = - Pij_reg;
```

Comment:
Be consistent. You (and other codes currently in Mjolnir) don't have any space between `operator-` and its operand. But here you have one.
```suggestion
    const coordinate_type Pji_reg       = -Pij_reg;
```